xplique/example_based/datasets_operations/tf_dataset_operations.py

```diff
@@ -0,0 +1,292 @@
+"""
+Set of functions to manipulated `tf.data.Dataset`
+"""
+from itertools import product
+
+import numpy as np
+import tensorflow as tf
+
+from ...types import Optional, Union
+
+
+def _almost_equal(arr1, arr2, epsilon=1e-6):
+    """Ensure two array are almost equal at an epsilon"""
+    return np.shape(arr1) == np.shape(arr2) and np.sum(np.abs(arr1 - arr2)) < epsilon
+
+
+def are_dataset_first_elems_equal(dataset1: Optional[tf.data.Dataset] = None,
+                                  dataset2: Optional[tf.data.Dataset] = None,
+                                  ) -> bool:
+    """
+    Test if the first batch of elements of two datasets are the same.
+    It is used to verify equality between datasets in a lazy way.
+
+    Parameters
+    ----------
+    dataset1
+        First `tf.data.Dataset` to compare.
+    dataset2
+        Second `tf.data.Dataset` to compare.
+
+    Returns
+    -------
+    test_result
+        Boolean value of the equality.
+    """
+    if dataset1 is None:
+        return dataset2 is None
+
+    if dataset2 is None:
+        return False
+
+    next1 = next(iter(dataset1))
+    next2 = next(iter(dataset2))
+    if isinstance(next1, tuple):
+        next1 = next1[0]
+        if isinstance(next2, tuple):
+            next2 = next2[0]
+        else:
+            return False
+
+    return _almost_equal(next1, next2)
+
+
+def is_batched(dataset: tf.data.Dataset) -> bool:
+    """
+    Check if a TensorFlow dataset is batched.
+
+    Parameters
+    ----------
+    dataset : tf.data.Dataset
+        The dataset to check.
+
+    Returns
+    -------
+    bool
+        True if the dataset is batched, False otherwise.
+    """
+    # Extract the element_spec
+    spec = dataset.element_spec
+
+    # Handle datasets with tuple or dict structures
+    if isinstance(spec, (tuple, dict)):
+        # Check if any part of the element_spec is batched
+        if isinstance(spec, tuple):
+            return all(s.shape[0] is None for s in spec)
+        if isinstance(spec, dict):
+            return all(s.shape[0] is None for s in spec.values())
+    else:
+        # Check if the first dimension is None (indicating batching)
+        return spec.shape[0] is None
+
+    # If we reach here, it's not batched
+    return False
+
+
+def is_not_shuffled(dataset: Optional[tf.data.Dataset]) -> bool:
```

Comment:
The name is not really well suited. In addition, I would rather use an iterator to check that is two instantiations of the same dataset..

Reply:
I agree on the name, but what do you mean on the iterator?